arch/x86/64/include/segment-definitions.h

```diff
@@ -0,0 +1,134 @@
+#define D_CONFORMING  0x04 // For code segments only
+
+// General segment descriptor type flags
+#define D_DPL0 0x00
+#define D_DPL1 0x20
+#define D_DPL2 0x40
+#define D_DPL3 0x60
+
+#define D_PRESENT 0x80
+
+#define D_64BIT     0x2000
+#define D_SIZE      0x4000
+#define D_G_PAGE    0x8000
+
+
+
+typedef struct
+{
+    uint16 rpl     : 2;
+    uint16 ldt     : 1; // Use local descriptor table
+    uint16 offset  :13;
+}__attribute__((__packed__)) SegmentSelector;
+
+
+typedef struct
+{
+    uint16 limitL;           // Limit is ignored for data segments in long mode
+    uint16 baseLL;
+    uint8 baseLM;
+    uint8 accessed       :1;
+    uint8 writeable      :1;
+    uint8 direction      :1; // 0 = up, 1 = down
+    uint8 is_code_seg    :1; // 0 = data segment, 1 = code segment
+    uint8 is_normal_seg  :1; // 0 = data/code segment, 1 = system segment
+    uint8 dpl            :2;
+    uint8 present        :1;
+    uint8 limitH         :4;
+    uint8 ignored        :1;
+    uint8 zero           :1;
+    uint8 stack_seg_size :1; // Size of expand-down stack segments
+    uint8 granularity    :1; // 0 = limit is byte sized, 1 = limit is page (4k) sized
+    uint8 baseLH;
+    uint32 baseH;            // Only actually exists in TSS and LDT descriptors in long mode
+    uint32 reserved;         // Only actually exists in TSS and LDT descriptors in long mode
+}__attribute__((__packed__)) DataSegmentDescriptor;
+
+typedef struct
+{
+    uint16 limitL;          // Limit is ignored for code segments in long mode
+    uint16 baseLL;
+    uint8 baseLM;
+    uint8 accessed      :1;
+    uint8 readable      :1;
+    uint8 conforming    :1;
+    uint8 is_code_seg   :1; // 0 = data segment, 1 = code segment
+    uint8 is_normal_seg :1; // 0 = data/code segment, 1 = system segment
+    uint8 dpl           :2;
+    uint8 present       :1;
+    uint8 limitH        :4;
+    uint8 ignored       :1;
+    uint8 long_mode     :1; // 64-bit flag (0 = compatibility mode, 1 = normal 64-bit mode)
+    uint8 addr_size     :1; // Default data/address size in compatibility mode. (0 = 16 bit, 1 = 32 bit). Must be 0 when long_mode = 1
+    uint8 granularity   :1;
+    uint8 baseLH;
+    uint32 baseH;           // Only actually exists in TSS and LDT descriptors in long mode
+    uint32 reserved;        // Only actually exists in TSS and LDT descriptors in long mode
+}__attribute__((__packed__)) CodeSegmentDescriptor;
+
+typedef struct
+{
+    uint16 limitL;
+    uint16 baseLL;
+    uint8 baseLM;
+    uint8 type          :4; // system segment type (TSS, LDT, ...)
+    uint8 is_normal_seg :1; // 0 = system segment, 1 = data/code segment
+    uint8 dpl           :2;
+    uint8 present       :1;
+    uint8 limitH        :4;
+    uint8 ignored       :1;
+    uint8 zero          :2;
+    uint8 granularity   :1;
+    uint8 baseLH;
+    uint32 baseH;
+    uint32 reserved;
+}__attribute__((__packed__)) SystemSegmentDescriptor;
+
+typedef union
+{
+    DataSegmentDescriptor dataseg;
```

Comment:
i can't find where they are used, are they unused?

Reply:
Base SWEB doesn't do much with segments aside from initializing them, so they aren't really used anywhere by default. It's more for future use and to replace the opaque `typeL`/`typeH`in the old definition.
setSegmentDescriptor() could use them, but it's not really necessary there.